cabal-testsuite/src/Test/Cabal/NeedleHaystack.hs

```diff
@@ -0,0 +1,183 @@
+            txFwd :: (String -> String)
+        }
+
+-- | Identity transformation for the search strings and the text to search in,
+-- leaves them unchanged.
+txContainsId :: TxContains
+txContainsId = TxContains id id
+
+-- | Conversions of the needle and haystack strings, the seach string and the
+-- text to search in.
+data NeedleHaystack =
+    NeedleHaystack
+        {
+            expectNeedleInHaystack :: Bool,
+            displayHaystack :: Bool,
+            txNeedle :: TxContains,
+            txHaystack :: TxContains
+        }
+
+-- | Symmetric needle and haystack functions, the same conversion for each going
+-- forward and the same coversion for each going backward.
+symNeedleHaystack :: (String -> String) -> (String -> String) -> NeedleHaystack
+symNeedleHaystack bwd fwd = let tx = TxContains bwd fwd in NeedleHaystack True False tx tx
+
+-- | Multiline needle and haystack functions with symmetric conversions. Going
+-- forward converts line breaks to @"\\n"@.  Going backward adds visible
+-- delimiters to lines.
+multilineNeedleHaystack :: NeedleHaystack
+multilineNeedleHaystack = symNeedleHaystack delimitLines encodeLf
+
+-- | Minimal set up for finding the needle in the haystack. Doesn't change the
+-- strings and doesn't displaying the haystack in any assertion failure message.
+needleHaystack :: NeedleHaystack
+needleHaystack = NeedleHaystack True False txContainsId txContainsId
+
+-- | Replace line breaks with spaces, correctly handling @"\\r\\n"@.
+--
+-- >>> lineBreaksToSpaces "foo\nbar\r\nbaz"
+-- "foo bar baz"
+--
+-- >>> lineBreaksToSpaces "foo\nbar\r\nbaz\n"
+-- "foo bar baz"
+--
+-- >>> lineBreaksToSpaces "\nfoo\nbar\r\nbaz\n"
+-- " foo bar baz"
+lineBreaksToSpaces :: String -> String
+lineBreaksToSpaces = unwords . lines . filter ((/=) '\r')
+
+-- | Replaces path separators found with those of the current OS, URL-like paths
+-- excluded.
+--
+-- > buildOS == Linux; normalizePathSeparators "foo\bar\baz" => "foo/bar/baz"
+-- > buildOS == Windows; normalizePathSeparators "foo/bar/baz" => "foo\bar\baz"
+normalizePathSeparators :: String -> String
+normalizePathSeparators =
+    unlines . map normalizePathSeparator . lines
+    where
+        normalizePathSeparator p =
+            if | any (isJust . parseURI) (tails p) -> p
+               | buildOS == Windows ->
+                    [if Posix.isPathSeparator c then Windows.pathSeparator else c| c <- p]
+               | otherwise ->
+                    [if Windows.isPathSeparator c then Posix.pathSeparator else c| c <- p]
+
+-- | @unlines@ from base will add a trailing newline if there isn't one already
+-- but this one doesn't
+--
+-- >>> lines "abc"
+-- ["abc"]
+--
+-- >>> Data.List.unlines $ lines "abc"
+-- "abc\n"
+--
+-- >>> unlines $ lines "abc"
+-- "abc"
+unlines :: [String] -> String
+unlines = maybe "" fst . unsnoc . Prelude.unlines
+
+-- | @unsnoc@ is only in base >= 4.19 so we copy its definition here rather than
```

Comment:
Maybe this belongs in a `Distribution.Compat` module?

Reply:
Thanks @geekosaur. I should have looked harder. I found an `unsnoc` in `Cabal-syntax`, so switched to using that one:

https://github.com/haskell/cabal/blob/5ad146cd1aa28c6d34be124ab41352fa1878771f/Cabal-syntax/src/Distribution/Utils/Generic.hs#L531-L533

Like `unsnoc` in base, it uses lazy pattern matching but is implemented using `unsnocNE`.

Reply:
Actually I'd have the same question there, but maybe that makes it a more intrusive change requiring its own PR.